src/neo/SmartContract/ApplicationEngine.cs

```diff
@@ -259,11 +261,31 @@ internal ContractTask<T> CallFromNativeContract<T>(UInt160 callingScriptHash, UI
         protected override void ContextUnloaded(ExecutionContext context)
         {
             base.ContextUnloaded(context);
+            if (context.Script != CurrentContext?.Script)
+            {
+                ExecutionContextState state = context.GetState<ExecutionContextState>();
+                if (UncaughtException is null)
+                {
+                    state.Snapshot?.Commit();
+                    if (CurrentContext != null)
+                    {
+                        ExecutionContextState contextState = CurrentContext.GetState<ExecutionContextState>();
+                        contextState.NotificationCount += state.NotificationCount;
+                    }
+                }
+                else
+                {
+                    if (state.NotificationCount > 0)
+                        notifications.RemoveRange(notifications.Count - state.NotificationCount, state.NotificationCount);
```

Comment:
BTW, it looks to me that the notification items will still be counted in the refcounter this way, which is somewhat unexpected (they're technically gone).

Reply:
But if we don't count them then we can try to attack an RPC node with diagnostics enabled (that has to save these notifications). From DoS perspective it'd be the same for storage changes, diagnostic resource requirements can now significantly differ from regular execution (unless limited in some way).

Reply:
> BTW, it looks to me that the notification items will still be counted in the refcounter this way, which is somewhat unexpected (they're technically gone).

Once these notifications are consumed by the syscall, they are not reference counted unless the objects have other references on the stack.

> From DoS perspective it'd be the same for storage changes, diagnostic resource requirements can now significantly differ from regular execution (unless limited in some way).

It makes no difference. You need to pay for the syscalls, and the gas paid will not be refunded if an exception is thrown.

Reply:
> Once these notifications are consumed by the syscall, they are not reference counted unless the objects have other references on the stack.

But we count ones emitted from native contracts like in #2731? So we count these notification items, but not count ones emitted via `System.Runtime.Notify`? Why is that? I'd expect them to be treated equally.

> It makes no difference. You need to pay for the syscalls

Yeah, that's true. The only problem could be creating exceeding item count via notifications.

Reply:
#2731 Fixed an issue where reference counters did not work correctly if a contract tried to call `System.Runtime.GetNotifications` to get those notifications after native contracts sent them. But that doesn't mean that notifications issued take up reference counts. So the behavior of the two is the same.

Reply:
> But that doesn't mean that notifications issued take up reference counts

Then I've misinterpreted #2731, thanks for clarifying.

Reply:
> But that doesn't mean that notifications issued take up reference counts. So the behavior of the two is the same.

Just to make sure I'm not missing anything else, how that works wrt https://github.com/neo-project/neo-vm/blob/master/src/neo-vm/Types/Array.cs#L70 ? To me it looks like internal elements are added immediately.

Reply:
We add the references of the elements first. If the `Array` is not pushed onto the stack, then `ReferenceCounter.CheckZeroReferred()` will clear these references.